src/core/unit.c

```diff
@@ -303,16 +303,14 @@ bool unit_may_gc(Unit *u) {
 
         state = unit_active_state(u);
 
-        /* If the unit is inactive and failed and no job is queued for
-         * it, then release its runtime resources */
+        /* If the unit is inactive and failed and no job is queued for it, then release its runtime resources */
         if (UNIT_IS_INACTIVE_OR_FAILED(state) &&
             UNIT_VTABLE(u)->release_resources)
                 UNIT_VTABLE(u)->release_resources(u);
 
-        /* But we keep the unit object around for longer when it is
-         * referenced or configured to not be gc'ed */
+        /* But we keep the unit object around for longer when it is referenced or configured to not be gc'ed */
         if (state != UNIT_INACTIVE)
-                return false;
+                return true;
```

Comment:
This **MUST** remain `false`, because the output values of this function were inverted by commit https://github.com/systemd-rhel/rhel-7/commit/1716821fc5b93667a0bf821b25905de00818aec9#diff-a89a9b6f80aada989d298b4c2c3a9d64L285 which is **NOT** in RHEL 7.4, that's why here the patch says `true` (https://github.com/systemd-rhel/rhel-7/pull/49/commits/3aaaa276e24588bfeced6c739ec03babb50f6870#diff-a89a9b6f80aada989d298b4c2c3a9d64R308).

Also, see the comment at the beginning of this function - we wan't to keep units which are not in the `UNIT_INACTIVE` state:

```
        /* Checks whether the unit is ready to be unloaded for garbage collection.
         * Returns true when the unit may be collected, and false if there's some
         * reason to keep it loaded.
         *
         * References from other units are *not* checked here. Instead, this is done
         * in unit_gc_sweep(), but using markers to properly collect dependency loops.
         */
```

Reply:
This was a sloppy handling of diff files on my side. Thanks for catching this!